be/src/exec/table_function_node.cpp

```diff
@@ -0,0 +1,341 @@
+    RETURN_IF_ERROR(exec_debug_action(TExecNodePhase::GETNEXT));
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+
+    const RowDescriptor& parent_rowdesc = row_batch->row_desc();
+    const RowDescriptor& child_rowdesc = _children[0]->row_desc();
+    if (_parent_tuple_desc_size == -1) {
+        _parent_tuple_desc_size = parent_rowdesc.tuple_descriptors().size();
+        _child_tuple_desc_size = child_rowdesc.tuple_descriptors().size();
+        for (int i = 0; i < _child_tuple_desc_size; ++i) {
+            _child_slot_sizes.push_back(child_rowdesc.tuple_descriptors()[i]->slots().size());
+        }    
+    }
+
+    uint8_t* tuple_buffer = nullptr;
+    Tuple* tuple_ptr = nullptr;
+    Tuple* pre_tuple_ptr = nullptr;
+    int row_idx = 0;
+
+    while (true) {
+        RETURN_IF_CANCELLED(state);
+        RETURN_IF_ERROR(state->check_query_state("TableFunctionNode, while getting next batch."));
+
+        if (_cur_child_batch == nullptr) {
+            _cur_child_batch.reset(new RowBatch(child_rowdesc, state->batch_size(), mem_tracker().get()));
+        }
+        if (_child_batch_exhausted) {
+            // current child batch is exhausted, get next batch from child
+            RETURN_IF_ERROR(_children[0]->get_next(state, _cur_child_batch.get(), eos));
+            if (*eos) {
+                break;
+            }
+            _cur_child_offset = 0;
+            RETURN_IF_ERROR(_process_next_child_row());
+            if (_child_batch_exhausted) {
+                _cur_child_batch->reset();
+                continue;
+            }
+        }
+
+        while (true) {
+            int idx = _find_last_fn_eos_idx();
+            if (idx == 0) {
+                // all table functions' results are exhausted, process next child row
+                RETURN_IF_ERROR(_process_next_child_row());
+                if (_child_batch_exhausted) {
+                    break;
+                }
+            } else if (idx < _fn_num && idx != -1) {
+                // some of table functions' results are exhausted
+                if (!_roll_table_functions(idx)) {
+                    // continue to process next child row
+                    continue; 
+                }
+            }
+
+            // get slots from every table function
+            // Notice that _fn_values[i] may be null if the table function has empty result set.
+            for (int i = 0; i < _fn_num; i++) {
+                RETURN_IF_ERROR(_fns[i]->get_value(&_fn_values[i]));
+            }
+
+            // allocate memory for row batch for the first time
+            if (tuple_buffer == nullptr) {
+                int64_t tuple_buffer_size;
+                RETURN_IF_ERROR(
+                        row_batch->resize_and_allocate_tuple_buffer(state, &tuple_buffer_size, &tuple_buffer));
+                tuple_ptr = reinterpret_cast<Tuple*>(tuple_buffer);
+            }
+            
+            pre_tuple_ptr = tuple_ptr;
```

Comment:
The logic of constructing a new line can be abstracted separately into a function

Reply:
I tried. But it has to pass a lot param to that new method, which seems not necessary.